2.4.0 Filter/Subfilter Overall, tmdb_director, and tmdb_writer

meisnate12 opened the pull request:
## [2.4.0] - 2020-11-03
### Added
- #63 - Added `plex_search` to AND searches together
- Added additional `filters` and allow for use of `.not` for inverse `filters`
- #75 - Added `tmdb_director` and `tmdb_writer` which function the same as `tmdb_actor` but for directors and writers
- More compatibility  with previous config files.

### Changed
- #63 - `Plex Filters` are now listed and have been tested and have been changed to `Plex Searches`
- `subfilters` are now listed and have been tested and have been changed to `filters`

### Fixed
- `collection_order` was in the code as `collection_sort`
- Upgrade PlexAPI dependency to 4.2.0

mza921 commented on app/plex_tools.py:
```diff
@@ -178,43 +206,72 @@ def add_to_collection(config_path, plex, method, value, c, map, subfilters=None)
+                    elif method[-4] == ".gte":
+                        gte = True
+                        method = method[:-4]
+                    if method == "days_from_now":
```
Can we call this `max-age`, with days being the unit? Even nicer (but not necessary), support values like `30d` or `1y`?

meisnate12 replied:
absolutely lol `max_age` is way better than `days_from_now` and ill have it default to days with a straight int but if it ends with the `y` we can do years

mza921 replied:
I say keep it simple... just support days

mza921 commented on app/trakt_tools.py:
```diff
@@ -1,6 +1,7 @@
 import config_tools
 from urllib.parse import urlparse
 import plex_tools
+from imdb_tools import regex_first_int
```
I think it's odd for this method to come from `imdb_tools`, instead of maybe a `misc` module.

meisnate12 replied:
I mean I agree tbh but I wasn't trying to make a whole new file for one method but I'm sure there are some others we could move around. I'd prefer fewer files in general. I'd probably have four files `config_tools`, `plex_auto_collection`, `plex_tools`, and `api_tools` maybe even not have plex tools vs api tools and just have one file for all the connections to other services. In the end, this is a conversation for another day lol

mza921 commented on app/plex_tools.py:
```diff
@@ -178,43 +206,72 @@ def add_to_collection(config_path, plex, method, value, c, map, subfilters=None)
-                        break
-                if match:
+                        # Get the intersection of the user's terms and movie's terms
+                        # If it's empty and negitive is false, it's not a match
```
"negative"

mza921 commented on README.md:
```diff
@@ -436,16 +522,35 @@ Note that if you have multiple movie Libraries or multiple show Libraries Tautul
+- `subtitle_language` (Matches every movie with the specified subtitle language) (Movie libraries only)
+
+#### Special Subfilter options (These options can only take one value each)
+- `days_from_now` (Matches any movie/show whos Originally Available date is within the last X days where X is the number you give as the value)
```
"whose"

mza921 commented on app/plex_auto_collections.py:
```diff
@@ -64,12 +151,58 @@ def update_from_config(config_path, plex, headless=False, no_meta=False, no_imag
+                        print("| Config Error: {} subfilter not supported".format(sf))
+            else:
+                print("| Config Error: subfilters attribute is blank")
+        and_filters = []
```
I'm not in love with the name `and_filter`, but this is good stuff.

burkasaurusrex replied:
So, just taking a step back - I _really_ like the functionality here, but I'm afraid it might be too confusing. What if we simplified it to just two attributes:
1. "Simple" Plex filters the way they are now - `genre`, `studio`, etc., where multiple filter `or` together simply. I'd be fine if we wanted to create an attribute `plex_filter` - seems like it might simplify the code too.
2. "Custom" filters, where there is an attribute `custom_filter` that is an ordered dictionary or filters and `custom_filter_logic` that can rearrange the filters. I'd be fine if we call this `custom_plex_filter` and `custom_plex_filter_logic` too - seems like it might be more readable. For example

```yaml
Good Romance or Romantic Comedies:
  custom_filter:
    - genre: romance               # 1
    - genre: comedy                # 2
    - rating.gte: 7                # 3
  custom_filter_logic: ({1} or ({1} and {2})) and {3}
```

Feels like we could borrow Tautulli's method here to make this easy

https://github.com/Tautulli/Tautulli/blob/14b98a32e085d969f010f0249c3d2f660db50880/plexpy/helpers.py#L925-L1007

meisnate12 replied:
looking at this it looks really nice but I think it would be hard to combine subfilters and filters like that. Subfliters affect all movies in the collection including whats added via other lists/attributes where the original filters are just a one time search. Another problem with using `.and` is that we can't `.and.not`  so we would need another method thats maybe `.ant`. 

Lets not mix the subfilters with the filters so what we need to decide is what to do about `.and` personally my vote would be to use `.and` and `.ant` i like the commands being all the same length at the end

meisnate12 replied:
oh wait lets prefix it with `and.` and then it wont get in the way of not and we wont need another identifier

i like it, i think it looks nice like that

```yaml
collections:
  Dave Chappelle:
    actor: Dave Chappelle
    and.genre: Comedy
```

mza921 replied:
The documentation of filters vs. subfilters is great and even more thorough than the `plexapi`, but I agree that it's still potentially confusing. I think we should stick to the original structure until we can nail down how to simplify it to something like @burkasaurusrex posted with "simple" and "complex/custom/advanced" filters. This is closest to how it's presented in the Plex UI. It does mean that filters and subfilters are mixed and would complicate how things are passed to the `plexapi`. I think we should also normalize the filter names to whatever is in the Plex UI, such as using `release_date` vs. `originally_available`. Custom things like `max_age`, we can name ourselves. 

For now, let's just add these features:
- `max_age`
- `.gte` and `.lte`

burkasaurusrex commented on requirements.txt:
```diff
@@ -1,5 +1,5 @@
 # Remove
-PyYAML==5.1.2
+PyYAML==5.3.1
```
We should just phase out `PyYAML` altogether in favor of `raumel.yaml` to avoid two dependencies for similar functionality

meisnate12 replied:
where do we use PyYAML?

burkasaurusrex replied:
Actually, @mza921 it looks like you may have introduced `ruamel.yaml`? Is there a preference over `PyYAML`? I don't have an opinion, but we should standardize on one throughout the project I think.

mza921 replied:
I use `ruamel.yaml` for updating the config file during initial Trakt authorization. Research had led me to [this](https://stackoverflow.com/questions/28557626/how-to-update-yaml-file-using-python). I prefer it over `PyYAML` but didn't bother to update the other modules since I mostly worked on Trakt integration.

burkasaurusrex replied:
Ok, let's default to `ruamel.yaml` then and remove `PyYAML`. It looks like it's only used in a couple of places:

- https://github.com/mza921/Plex-Auto-Collections/blob/8d129ad590551ccd1656e07b85d308c8d0514908/app/config_tools.py#L80
- https://github.com/mza921/Plex-Auto-Collections/blob/8d129ad590551ccd1656e07b85d308c8d0514908/app/config_tools.py#L455
- https://github.com/mza921/Plex-Auto-Collections/blob/8d129ad590551ccd1656e07b85d308c8d0514908/app/radarr_tools.py#L1 (looks like this import can just be deleted)

meisnate12 commented:
Ok so I managed to add and. and .not as detailed in the readme. I wanna leave this functionality in but if we want to leave it out of the read me we can, but I think it's laid out well. 

On another note... subfilters and filters are functionally very very different. I don't think its going to be easy if even possible to actually combine them. Filters search using the LibrarySection search function which uses plex's search to do its work and its only for the one method. Subfilters are very very different. After any list type gets the movies its supposed to add to the collection each one is run aginst every subfilter checking the actual attributes of the movie object vs the subfilters attributes. I think we need to think of subfilters as a completely separate entity then filters because filters are limited to dealing with themselves whereas subfilters deal with every single list type. 

My newest code push changed the way the program reads the collection's attributes. before we would loop through every method and add all the method names except those in details to a list, then the program would loop through the list and use the dictionary to get the values, and finally, it would check those values once it got to the actual method to add them.

I changed this to where the first loop checks and makes sure the attribute entered is a valid attribute, is not blank, and validate the actually value when it can (the regex method has been moved to here) and then add the method and values to a paired list to be looped through afterward to actually process the attributes and their values which have all been validated at that point. I want to add more checks but that will be for a future release. 

I don't have any more huge sections of code I'm working on for this release but I changed a lot in my last commit so I'd like you both to review it all once again. 

P.S. the filter .not's are not currently working (subfilter .not works fine) the code is correct but the PlexAPI package accepts the ! after the attribute name but the filters only work for studio and year. I've opened an issue on their git about it [here](https://github.com/pkkid/python-plexapi/issues/595) so i say we take the not out of the readme for now but i think the and will be fine if they're confusing people can just not use them the simple filters still work

mza921 commented:
Naming things is [hard](https://www.karlton.org/2017/12/naming-things-hard/). Based on how subfilters work against the different list types, I would name them `global_filters` instead. This would also open the door to using filtering methods from other APIs. For example, I find that Trakt's `language` and `countries` filter is more accurate than Plex's, so I might want to incorporate a `trakt_filters` option in a future release. Same goes for IMDB or TMDB.

meisnate12 commented:
lmao accurate naming things is very hard I hate it so much. I don't really like `subfilters` either after going through all this code the name just makes it confusing and makes it seem like a subset of filters which it isn't. I like `global_filters` an alternative could be `list_filters` or another option is to change Plex filters to no longer use the word filter and call it plex search everywhere and make it so anytime the word filter is used in any way it's talking about the `subfilters`. 

The TMDb specific filters are not really filtering your movies based on TMDb. It just running the standard plex search for actor, director, or writer and getting the name pic and bio from TMDB. Basically TMDb is just doing the metadata for the collection while the plex search is what is actually adding the movies to your collections. So they're kind of special tags in that way becasue `tmdb_actor: 4169` is the same as doing `actor: Dave Chappelle`, `tmdb_profile: 4169`, and `tmdb_biography: 4169` but with less lines in the config. I don't think we should have them under a catagory called tmdb filters personally since it uses other things. 

I'll get the read me changed today and it looks like the next release of the plexapi will fix all the not problems were having

mza921 commented:
1. Yes, let's change `subfilters`
2. Yes, change Plex filters to search. Will users now have to specify `plex_search` as a list type or is it implied? 
```
collections:
  Dave Chappelle:
    plex_search:
      - actor: Dave Chappelle
```
3. Just use `filters`. Good point about `global_filters` being misleading, since these are scoped at the collection level after all list types are combined.

While I think these are the best way forward, they do break compatibility with existing config files. We need to make sure that the interactive mode also works with these changes.

meisnate12 commented:
So I'm thinking I'll be able to have it not break the old configs but it will instead give a warning for any attribute deprecated saying that it will be run as the proper attribute 

I hadn't thought about plex_search being a whole category but I'm not mad at it we could make everything under one search AND together and then multiple plex_searches OR together and take out the and. Completely

meisnate12 commented:
Also I don't think I've ever used interactive mode lol but I'll start testing it as well

meisnate12 commented:
ok introducing `plex_search` let me know what yall think?

meisnate12 commented:
Oh this version should work with any version of the config i put checks in for all the old attributes

mza921 commented on README.md:
```diff
@@ -1,6 +1,6 @@
-##### Version 2.2.1
-Plex Auto Collections is a Python 3 script that works off a configuration file to create/update Plex collections. Collection management with this tool can be automated in a varying degree of customizability. Supports IMDB, TMDb, and Trakt lists as well as built in Plex filters such as actors, genres, year, studio and more.
+##### Version 2.3.0
+Plex Auto Collections is a Python 3 script that works off a configuration file to create/update Plex collections. Collection management with this tool can be automated in a varying degree of customizability. Supports IMDB, TMDb, and Trakt lists as well as built in Plex Searchs using actors, genres, year, studio and more.
```
searches

mza921 commented on README.md:
```diff
@@ -146,10 +146,10 @@ Each collection is defined by the mapping name which becomes the name of the Ple
 
-The only required attribute for each collection is the list type. There are eleven different list types to choose from:
-- [Plex Collection](#plex-collection-list-type)
+The only required attribute for each collection is the list type. There are many different list types to choose from:
```
Happy not having to count any more

mza921 commented on config/config.yml.template:
```diff
@@ -33,10 +33,12 @@ collections:
-    year: [1990,1991,1992,1993,1994,1995,1996,1997,1998,1999]
-    summary: A collection of 1990s movies
+    collection_order: alpha
+  90s Comedys:
```
Comedies

mza921 reviewed (commented):
Thanks for all the updates.

meisnate12 commented:
no problem I like this script and want to see it improved plus learning python in the process is a great bonus

Btw just push your changes for the new movie Agent and ill get this pull rebased on them the same day so this can go out shortly after

mza921 commented on README.md:
```diff
@@ -173,12 +173,11 @@ You can create a collection based on the Plex search feature using the `plex_sea
 - `genre` (Gets every movie/show with the specified genre)
 - `studio` (Gets every movie/show with the specified studio)
-- `year` (Gets every movie/show with the specified year)
+- `year` (Gets every movie/show with the specified year) (Put a `-` between two years for a range i.e. `year: 1990-1999` or end with `NOW` to go till current i.e. `year: 2000-NOW`)
```
Can we keep undocumented support for `decade`?

meisnate12 replied:
well i took out decade in favor of the year range i added just do `year: 1990-1999` i can add it back in if needed but i figured the range would be less confusing and not bog down the user with another option plus the range isn't limited to 10 years either

meisnate12 replied:
Did you want me to add decade back or is year range ok?

mza921 replied:
I prefer to keep it

meisnate12 replied:
ok ill add it back silently

meisnate12 commented:
@mza921 my pull is now rebased on 2.3.0 and i upgraded my release to 2.4.0

mza921 commented on CHANGELOG.md:
```diff
@@ -4,7 +4,22 @@ All notable changes to this project will be documented in this file.
-## [2.3.0] - 2020-11-02
+## [2.4.0] - 2020-11-03
+### Added
+- Added `plex_search` to AND searches together
```
I'm confused about this. So if `plex_search` does an AND, how do I do an OR? How would this README example work:
```
collections:
  90s Movies:
    plex_search:
      year:
        - 1990
        - 1991
        - 1992
        - 1993
        - 1994
        - 1995
        - 1996
        - 1997
        - 1998
        - 1999
```

meisnate12 replied:
it ANDs different attributes and ORs multiples of the same attribute I can try and clarify that in the readme. The program will also tell you exactly how it's searching so for your example it would look like this:

```
| Processing Plex Search: year(1990 OR 1991 OR 1992 OR 1993 OR 1994 OR 1995 OR 1996 OR 1997 OR 1998 OR 1999)
``` 
it will show each AND on the subsequent lines

if you want to AND the same attribute you have to use subfilters

mza921 replied:
Got it. Nice addition to report the actual query.

mza921 commented:
Can you rebase on 2.3.1? Then we're good to go.

meisnate12 commented:
can do give me a minute

meisnate12 commented:
ok were good to go i also updated the changelog to include issue and pull numbers for easier tracking
